Sources/SwiftSyntax/SourceLocation.swift

```diff
@@ -145,13 +225,40 @@ public final class SourceLocationConverter {
     }
   }
 
-  /// Convert a ``AbsolutePosition`` to a ``SourceLocation``. If the position is
+  /// Convert a ``AbsolutePosition`` to a ``SourceLocation``.
+  ///
+  /// If the position is exceeding the file length then the ``SourceLocation``
+  /// for the end of file is returned. If position is negative the location for
+  /// start of file is returned.
+  public func location(for position: AbsolutePosition) -> SourceLocation {
+    let physicalLocation = physicalLocation(for: position)
+    if let lastSourceLocationDirective = sourceLocationDirectives.last(where: { $0.sourceLine < physicalLocation.line }),
+      let arguments = lastSourceLocationDirective.arguments
+    {
+      let presumedLine = arguments.line + physicalLocation.line - lastSourceLocationDirective.sourceLine - 1
+      return SourceLocation(
+        line: physicalLocation.line,
+        column: physicalLocation.column,
+        offset: physicalLocation.offset,
+        file: physicalLocation.file,
+        presumedLine: presumedLine,
+        presumedFile: arguments.file
+      )
+    }
+
+    return physicalLocation
+  }
+
+  /// Compute the location of `position` without taking `#sourceLocation`
+  /// directives into account.
+  ///
+  /// If the position is
   /// exceeding the file length then the ``SourceLocation`` for the end of file
   /// is returned. If position is negative the location for start of file is
   /// returned.
```

Comment:
Why do we allow negative `AbsolutePosition`?

Reply:
I don’t think there’s a real use case. Would you prefer to `assert` or `precondition`? 🤔

Reply:
`precondition` seems fine to me